src/NHibernate/Transaction/AdoNetWithDistributedTransactionFactory.cs

```diff
@@ -94,90 +79,196 @@ public class DistributedTransactionContext : ITransactionContext, IEnlistmentNot
 								{
-									logger.Debug(string.Format("[session-id={0}] Flushing from Dtc Transaction", sessionImplementor.SessionId));
-									sessionImplementor.Flush();
+									_logger.DebugFormat("[session-id={0}] Flushing from Dtc Transaction", _sessionImplementor.SessionId);
+									_sessionImplementor.Flush();
 								}
 							}
-							logger.Debug("prepared for DTC transaction");
+							_logger.Debug("prepared for DTC transaction");
 
 							tx.Complete();
 						}
+						// Lock the session to ensure second phase gets done before the session is used by code following
+						// the transaction scope disposal.
+						_waitEvent.Reset();
+
 						preparingEnlistment.Prepared();
 					}
 					catch (Exception exception)
 					{
-						logger.Error("DTC transaction prepare phase failed", exception);
+						_logger.Error("DTC transaction prepare phase failed", exception);
 						preparingEnlistment.ForceRollback(exception);
 					}
 				}
 			}
 
 			void IEnlistmentNotification.Commit(Enlistment enlistment)
+				=> ProcessSecondPhase(enlistment, true);
+
+			void IEnlistmentNotification.Rollback(Enlistment enlistment)
+				=> ProcessSecondPhase(enlistment, false);
+
+			void IEnlistmentNotification.InDoubt(Enlistment enlistment)
+				=> ProcessSecondPhase(enlistment, null);
+
+			private void ProcessSecondPhase(Enlistment enlistment, bool? success)
 			{
-				using (new SessionIdLoggingContext(sessionImplementor.SessionId))
+				using (new SessionIdLoggingContext(_sessionImplementor.SessionId))
 				{
-					logger.Debug("committing DTC transaction");
-					// we have nothing to do here, since it is the actual
-					// DB connection that will commit the transaction
+					_logger.Debug(
+						success.HasValue
+							? success.Value
+								? "committing DTC transaction"
+								: "rolled back DTC transaction"
+							: "DTC transaction is in doubt");
+					// we have not much to do here, since it is the actual
+					// DB connection that will commit/rollback the transaction
+					// Usual cases will raise after transaction actions from TransactionCompleted event.
+					if (!success.HasValue)
+					{
+						// In-doubt. A durable ressource has failed and may recover, but we won't wait to know.
+						RunAfterTransactionActions(false);
+					}
+
 					enlistment.Done();
-					IsInActiveTransaction = false;
 				}
 			}
 
-			void IEnlistmentNotification.Rollback(Enlistment enlistment)
+			#endregion
+
+			public void TransactionCompleted(object sender, TransactionEventArgs e)
 			{
-				using (new SessionIdLoggingContext(sessionImplementor.SessionId))
+				e.Transaction.TransactionCompleted -= TransactionCompleted;
+				// This event may execute before second phase, so we cannot try to get the success from second phase.
+				// Using this event is required in case the prepare phase failed and called force rollback: no second
+				// phase would occur for this ressource.
```

Comment:
Documentation on which conditions the second phase will not be called at all lacks. At first I have put it all in second phase callbacks, just to realize thanks to some tests there were cases for which it was not good.

So the forced rollback from session volatile prepare phase is maybe only one case in which second phase do not occur. For staying on the safe side, we would better keep that handling in `TransactionCompleted`.

The fact the completion actions were called from in-doubt too made me wonder whether in in-doubt case, the transaction completed call be not called either. There too, for staying on the safe side, I have decided to keep that call. This transaction factory is the result of many years of fiddling with many failing case for trying to support them all. During my own fiddling with it, I have seen that many design decisions in it, lacking comments and which were looking as overly complicating the code, where indeed needed for some reasons. So I had to kept them, but at least adding comments for most of them.

Moreover, handling transaction completion actions from the prepare phase in case of force rollback may cause other issues, since the session connection state at that point is unknown: has it already processed its own prepare phase? Is it processing it concurrently? Raising completion actions there would trigger its release, and I do not know if all connection implementations would support that.

This is also why I consider we should still provide other transaction factories even if we appear to get this one working: there too many points where we still act on the connection with concurrency risks. Providing other transaction factories would allow whoever wishing it to not take those risks, at the cost of some more explicit flushes calls.